File: src/0234.palindrome-linked-list/main.c
// Copyright (c) 2023 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be
// found in the LICENSE file.

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

struct ListNode {
  int val;
  struct ListNode* next;
};

struct ListNode* listNew(int val) {
  struct ListNode* node = malloc(sizeof(struct ListNode));
  assert(node != NULL);
  node->val = val;
  node->next = NULL;
  return node;
}

struct ListNode* listFromArray(int arr[], int len) {
  struct ListNode* list = NULL;
  for (int i = len - 1; i >= 0; --i) {
    struct ListNode* node = listNew(arr[i]);
    node->next = list;
    list = node;
  }
  return list;
}

void listFree(struct ListNode* head) {
  while (head != NULL) {
    struct ListNode* node = head;
    head = head->next;
    free(node);
  }
}

void listDebug(struct ListNode* head) {
  while (head != NULL) {
    printf("%d, ", head->val);
    head = head->next;
  }
  printf("\n");
}

bool isPalindrome(struct ListNode* head) {
  if (head == NULL) {
    return true;
  }

  // 1. Reverse left half of list.
  struct ListNode* slow = head;
  struct ListNode* fast = head;
  struct ListNode* next = NULL;
  struct ListNode* prev = NULL;

  while (fast != NULL && fast->next != NULL) {
    slow = slow->next;
    fast = fast->next->next;

    // Reverse via slow node.
    next = head->next;
    head->next = prev;
    prev = head;
    head = next;
  }

  if (fast != NULL) {
    slow = slow->next;
  }
  head = prev;

  // 2. Compare from middle.
  while (slow != NULL) {
    if (slow->val != head->val) {
      return false;
    }
    head = head->next;
    slow = slow->next;
  }
  return true;
}

void checkSolution(int arr[], int len) {
  struct ListNode* list = listFromArray(arr, len);
  listDebug(list);
  printf("is palindrome: %d\n", isPalindrome(list));
  listFree(list);
}

int main() {
  int arr1[] = {1, 2, 2, 1};
  checkSolution(arr1, sizeof(arr1) / sizeof(arr1[0]));

  int arr2[] = {1, 2,};
  checkSolution(arr2, sizeof(arr2) / sizeof(arr2[0]));

  return 0;
}

File: src/0234.palindrome-linked-list/src/main.rs
// Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be found
// in the LICENSE file.

use std::ops::Deref;

/// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    #[must_use]
    pub const fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }

    #[must_use]
    pub fn from_slice(slice: &[i32]) -> Option<Box<Self>> {
        let mut list = None;
        for x in slice.iter().rev() {
            list = Some(Box::new(ListNode {
                val: *x,
                next: list,
            }));
        }
        list
    }
}

pub trait ListLinkTrait {
    fn next(self) -> Self;

    fn is_empty(&self) -> bool;

    fn len(&self) -> usize;

    fn value(&self) -> Option<i32>;
}

impl ListLinkTrait for Option<Box<ListNode>> {
    fn next(self) -> Self {
        match self {
            Some(boxed_self) => boxed_self.next,
            None => None,
        }
    }

    fn is_empty(&self) -> bool {
        self.is_none()
    }

    fn len(&self) -> usize {
        match self {
            Some(boxed_self) => boxed_self.deref().next.len() + 1,
            None => 0,
        }
    }

    fn value(&self) -> Option<i32> {
        self.as_ref().map(|head| head.val)
    }
}

pub type IsPalindromeFunc = fn(Option<Box<ListNode>>) -> bool;

/// Use vector as cache, iterator through linked list and compare values.
pub fn is_palindrome(head: Option<Box<ListNode>>) -> bool {
    let mut head = head;
    let len = head.len();
    if len == 0 {
        return false;
    }

    let mut cache = Vec::with_capacity(len);
    let mut index = 0;

    while let Some(head_ref) = head {
        cache.push(head_ref.val);
        head = head_ref.next;
        index += 1;
        if len / 2 == index {
            index -= 1;
            break;
        }
    }

    if len % 2 == 1 {
        head = head.next();
    }

    while let Some(head_ref) = head {
        if head_ref.val != cache[index] {
            return false;
        }
        if index == 0 {
            break;
        }
        index -= 1;
        head = head_ref.next;
    }

    true
}

/// Two pointers
///
/// Reverse left half of list and compare with right part.
pub fn is_palindrome2(head: Option<Box<ListNode>>) -> bool {
    if head.is_none() {
        return false;
    }

    let len = head.len();

    // 1. Reverse left half of list.
    let mut head = head;
    let mut index = 0;
    let mut left_part = None;
    while len / 2 != index {
        let val = unsafe { head.value().unwrap_unchecked() };
        left_part = Some(Box::new(ListNode {
            val,
            //val: head.value().unwrap(),
            next: left_part,
        }));
        head = head.next();
        index += 1;
    }

    if len % 2 == 1 {
        head = head.next();
    }

    // 2. compare from middle
    while head.is_some() {
        if head.value() != left_part.value() {
            return false;
        }
        head = head.next();
        left_part = left_part.next();
    }

    true
}

fn check_solution(func: IsPalindromeFunc) {
    let l1 = ListNode::from_slice(&[1, 2, 2, 1]);
    assert!(func(l1));

    let l1 = ListNode::from_slice(&[1, 2]);
    assert!(!func(l1));

    let l1 = ListNode::from_slice(&[1, 2, 1]);
    assert!(func(l1));
}

fn main() {
    check_solution(is_palindrome);
    check_solution(is_palindrome2);
}
